common/basecomponent/base.go

```diff
@@ -184,16 +256,46 @@ func (b *BaseDendrite) CreateKeyDB() keydb.Database {
 	if err != nil {
 		logrus.WithError(err).Panicf("failed to connect to keys db")
 	}
+	if b.LibP2P != nil {
+		mdns := mDNSListener{
+			host:  b.LibP2P,
+			keydb: db,
+		}
+		serv, err := p2pdisc.NewMdnsService(
+			b.LibP2PContext,
+			b.LibP2P,
+			time.Second*10,
+			"_matrix-dendrite-p2p._tcp",
+		)
+		if err != nil {
+			panic(err)
+		}
+		serv.RegisterNotifee(&mdns)
+	}
 
 	return db
 }
 
 // CreateFederationClient creates a new federation client. Should only be called
 // once per component.
 func (b *BaseDendrite) CreateFederationClient() *gomatrixserverlib.FederationClient {
-	return gomatrixserverlib.NewFederationClient(
-		b.Cfg.Matrix.ServerName, b.Cfg.Matrix.KeyID, b.Cfg.Matrix.PrivateKey,
-	)
+	if b.LibP2P != nil {
+		fmt.Println("Running in libp2p federation mode")
+		fmt.Println("Warning: Federation with non-libp2p homeservers will not work in this mode yet!")
+		tr := &http.Transport{}
+		tr.RegisterProtocol(
+			"matrix",
+			p2phttp.NewTransport(b.LibP2P, p2phttp.ProtocolOption("/matrix")),
+		)
+		return gomatrixserverlib.NewFederationClientWithTransport(
+			b.Cfg.Matrix.ServerName, b.Cfg.Matrix.KeyID, b.Cfg.Matrix.PrivateKey, tr,
+		)
```

Comment:
Again, just put this code in `main.go`, there's no reason to pollute this file with P2P stuff.